### ISSUE TITLE
Kubernetes Annotation based init_configs should be able to be stored in configmaps.

### ISSUE THREAD

**Issue by carsonoid** [DataDog/dd-agent#3518]
# The Problem

The Kubernetes annotation based Autodiscovery is awesome!  But when your init_config's get long it can be a nightmare to maintain them.  I recently discovered this when trying to monitor a K8s based kafka installation.

Here is an excerpt of my final annotations to give you an idea of how ugly it gets:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/created-by: |
      {"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"kafka-brittle","name":"kafka-broker-0-849680288","uid":"2facb78c-98c3-11e7-85f3-12d06627ddea","apiVersion":"extensions","resourceVersion":"4532762"}}
    service-discovery.datadoghq.com/broker.check_names: '["kafka"]'
    service-discovery.datadoghq.com/broker.instances: '[{"host":"%%host%%", "port":
      9999}]'
    service-discovery.datadoghq.com/broker.init_configs: |
      [
        {
          "is_jmx": true,
          "conf": [
            {
              "include": {
                "domain": "kafka.producer",
                "bean_regex": "kafka\\.producer:type=ProducerRequestMetrics,name=ProducerRequestRateAndTimeMs,clientId=.*",
                "attribute": {
                  "Count": {
                    "metric_type": "rate",
                    "alias": "kafka.producer.request_rate"
                  }
                }
              }
            },
           ....
          ]
        }
      ]
``` 

The base kafka config is almost 400 lines of non-minified json! I could minify it but then modifications become really difficult.

---

# The Suggestion

Add a new syntax for all the annotations which tells the dd-agent to read a specific config from a configmap.

### Sample Pod Annotation 

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
      annotations:
        service-discovery.datadoghq.com/broker.check_names: '["kafka"]'
        service-discovery.datadoghq.com/broker.init_configs: '["configmap:kafka:config"]'
        service-discovery.datadoghq.com/broker.instances: '[{"host":"%%host%%", "port": 9999}]'
```

### Sample ConfigMap JSON config

```yaml
apiVersion: v1
metadata:
  name: kafka
data:
  config: |
        [
          {
            "is_jmx": true,
            "conf": [
              {
                "include": {
                  "domain": "kafka.producer",
                  "bean_regex": "kafka\\.producer:type=ProducerRequestMetrics,name=ProducerRequestRateAndTimeMs,clientId=.*",
                  "attribute": {
                    "Count": {
                      "metric_type": "rate",
                      "alias": "kafka.producer.request_rate"
                    }
                  }
                }
              },
             ....
            ]
          }
        ]
```

### Sample ConfigMap YAML config

```yaml
apiVersion: v1
metadata:
  name: kafka
data:
  config: |
    is_jmx: true
    conf:
      - include:
          domain: 'kafka.producer'
          bean_regex: 'kafka\.producer:type=ProducerRequestMetrics,name=ProducerRequestRateAndTimeMs,clientId=.*'
          attribute:
            Count:
              metric_type: rate
              alias: kafka.producer.request_rate
      - include:
          domain: 'kafka.producer'
          bean_regex: 'kafka\.producer:type=ProducerRequestMetrics,name=ProducerRequestRateAndTimeMs,clientId=.*'
          attribute:
            Mean:
              metric_type: gauge
              alias: kafka.producer.request_latency_avg
      - include:
          domain: 'kafka.producer'
          bean_regex: 'kafka\.producer:type=ProducerTopicMetrics,name=BytesPerSec,clientId=.*'
          attribute:
            Count:
              metric_type: rate
              alias: kafka.producer.bytes_out
      - include:
          domain: 'kafka.producer'
          bean_regex: 'kafka\.producer:type=ProducerTopicMetrics,name=MessagesPerSec,clientId=.*'
          attribute:
            Count:
              metric_type: rate
              alias: kafka.producer.message_rate
```

Basically instead of giving the entire config dictionary in the list you just tell the agent to get the config from the config key of the kafka config map in the pod's namespace. The config should just be a long string of either json or yaml.

This takes the ugly config out of the pod's annotation and also allows you to re-use configurations if desired.

**Comment by xvello** [comment]
Hello @carsonoid ,

Thanks for your suggestion, but I believe that would add more complexity in debugging configuration issues. Here are the current options I see for your use case:

- Some users keep their config as yaml and have a home grown script that converts it to minified json before injecting it in their spec files. That modified spec is then sent to kubectl.
- You can store the full yaml autoconf template in a configmap and [mount it in](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#add-configmap-data-to-a-specific-path-in-the-volume) `/conf.d/auto_conf/kafka.conf`. The template will be used in [auto_conf file mode](https://docs.datadoghq.com/guides/autodiscovery/#template-source-files-auto-conf)

One alternative would be to write a new k/v store backed using kubernetes configmaps, as an alternative to our current [etcd](https://github.com/DataDog/dd-agent/blob/master/utils/service_discovery/etcd_config_store.py), consul & zookeper backend. Would you be interested in tackling that?


Regards

**Comment by carsonoid** [comment]
@xvello  Thanks for the follow up. The issue with mounting auto_conf files via config map is that The entire auto_conf directory get's mounted over. So the built-in auto_conf templates would be lost. I guess the external script is option is the best one right now.

**Comment by xvello** [comment]
Hi @carsonoid ,

If you mount a `/conf.d` folder inside the container, the entrypoint will copy the files to `/etc/dd-agent/conf.d` and preserve the existing ones if not overriden, see https://github.com/DataDog/docker-dd-agent/blob/12.1.5171/README.md#configuration-files


Regards
